### PR TITLE
[Form] Add parameter type declarations

### PR DESCRIPTION
**Pull request by vudaltsov**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #32179 
| License       | MIT
| Doc PR        | n/a

Some hints added, will finish within 2 days.

### CODE REVIEW

File: src/Symfony/Component/Form/ButtonBuilder.php [hunk 1]
```diff
@@ -167,7 +167,7 @@ public function getForm()
      *
      * @throws BadMethodCallException
      */
-    public function addEventListener($eventName, $listener, $priority = 0)
+    public function addEventListener(string $eventName, $listener, int $priority = 0)
```
Comment (yceruto) [inline src/Symfony/Component/Form/ButtonBuilder.php]: `callable $listener`?

File: src/Symfony/Bridge/Twig/Form/TwigRendererEngine.php [hunk 2]
```diff
@@ -40,7 +40,7 @@ public function __construct(array $defaultThemes, Environment $environment)
     /**
      * {@inheritdoc}
      */
-    public function renderBlock(FormView $view, $resource, $blockName, array $variables = [])
+    public function renderBlock(FormView $view, $resource, string $blockName, array $variables = [])
```
Comment (derrabus) [inline src/Symfony/Bridge/Twig/Form/TwigRendererEngine.php]: Careful: If you do this, TwigBridge 5.0 will become incompatible with Form 4.4! I would recommend to revert the changes you‘ve made to the bridge.

File: src/Symfony/Component/Form/ChoiceList/Factory/ChoiceListFactoryInterface.php [hunk 3]
```diff
@@ -31,13 +31,12 @@ interface ChoiceListFactoryInterface
...
-     * @param iterable      $choices The choices
-     * @param callable|null $value   The callable generating the choice
-     *                               values
+     * @param callable|null $value The callable generating the choice
+     *                             values
      *
      * @return ChoiceListInterface The choice list
      */
-    public function createListFromChoices($choices, $value = null);
+    public function createListFromChoices(iterable $choices, $value = null);
```
Comment (yceruto) [inline src/Symfony/Component/Form/ChoiceList/Factory/ChoiceListFactoryInterface.php]: `callable $value = null`?

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php [hunk 4]
```diff
@@ -40,7 +40,7 @@ public function __construct(callable $callback)
     /**
      * {@inheritdoc}
      */
-    public function loadChoiceList($value = null)
+    public function loadChoiceList(?callable $value = null)
```
Comment (yceruto) [inline src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php]: I think all nullable types can be removed if they already have `null` as default

File: src/Symfony/Bridge/Twig/Form/TwigRendererEngine.php (hunk 2 above)
Comment (vudaltsov) [reply src/Symfony/Bridge/Twig/Form/TwigRendererEngine.php]: @derrabus , why not bumping up bridge's minimum requirement for form via the `conflict` section?

Comment (derrabus) [reply src/Symfony/Bridge/Twig/Form/TwigRendererEngine.php]: I would not bump the requirements just for having type-hints. We could still do this for 5.1 – there‘s no rush here (imho).

File: src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php (hunk 4 above)
Comment (derrabus) [reply src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php]: Yes, that would be consistent with the rest of the codebase at least.
```suggestion
    public function loadChoiceList(callable $value = null)
```

File: src/Symfony/Component/Form/ButtonBuilder.php (hunk 1 above)
Comment (vudaltsov) [reply src/Symfony/Component/Form/ButtonBuilder.php]: @yceruto , this is to avoid autoloading. See `EventDispatcherInterface`

File: src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php (hunk 4 above)
Comment (vudaltsov) [reply src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php]: @yceruto , this is debatable. I cannot find the proof but I heard that relying on default nulls without adding `?` will be deprecated since it's an inconsistency.

Comment (vudaltsov) [reply src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php]: If we are talking about the current Symfony codebase consistency, then I agree

Comment (vudaltsov) [reply src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php]: Yes, we should remove `?` in this case for now. See https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3883#issue-338880999

File: src/Symfony/Component/Form/ButtonBuilder.php (hunk 1 above)
Comment (derrabus) [reply src/Symfony/Component/Form/ButtonBuilder.php]: On `EventDispatcherInterface`, this was done because (depending on how you use the dispatcher) you could potentially add 8000 listeners to 5687 different events and only call 50 of them during one request which leaves you with a lot of unnecessary autoload calls.

On forms however, it might be save to add the type-hint.

Comment (vudaltsov) [reply src/Symfony/Component/Form/ButtonBuilder.php]: We can add a `callable` typehint if the it is called sooner or later like in `*ChoiceLoader`. Event listeners are often never called.

Comment (vudaltsov) [reply src/Symfony/Component/Form/ButtonBuilder.php]: Well, in forms `*_SUBMIT` events are fired less often. So I think it makes sense to do same here.

File: src/Symfony/Component/Form/ButtonBuilder.php [hunk 5]
```diff
@@ -85,13 +85,13 @@ public function add($child, $type = null, array $options = [])
      *
      * This method should not be invoked.
      *
-     * @param string                   $name
-     * @param string|FormTypeInterface $type
```
Comment (vudaltsov) [inline src/Symfony/Component/Form/ButtonBuilder.php]: I removed `FormTypeInterface` because it was not supported by the factory anyway. The test was working with a mock factory and was not fully checking this case.

File: src/Symfony/Component/Form/ButtonBuilder.php (hunk 1 above)
Comment (vudaltsov) [reply src/Symfony/Component/Form/ButtonBuilder.php]: On the other hand, in most cases `*_SUBMIT` listeners are declared in subscribers together with `*_DATA` events which are fired for sure... @xabbuh , could you help us here?

File: src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php (hunk 4 above)
Comment (derrabus) [reply src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php]: > I heard that relying on default nulls without adding `?` will be deprecated since it's an inconsistency.

I doubt it because that would create strange edge-cases for applications that aim at broad compatibility ranges. But if that should ever happen, this can be fixed by a tool. Until then, we should aim for consistency.

File: src/Symfony/Component/Form/Button.php [hunk 6]
```diff
@@ -156,7 +154,7 @@ public function get($name)
     /**
      * Unsupported method.
      *
-     * @param string $name
+     * @param string|int $name
      *
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Button.php]: Docblock changes should usually go to lower branches. Please take care of this when reviewing.

File: src/Symfony/Component/Form/ButtonBuilder.php [hunk 7]
```diff
@@ -70,12 +70,12 @@ public function __construct(?string $name, array $options = [])
      * This method should not be invoked.
      *
      * @param string|int|FormBuilderInterface $child
-     * @param string|FormTypeInterface        $type
+     * @param string|null                     $type
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/ButtonBuilder.php]: For lower branches if legit.

File: src/Symfony/Component/Form/Button.php [hunk 8]
```diff
@@ -170,7 +168,7 @@ public function has($name)
...
      * This method should not be invoked.
      *
-     * @param string $name
+     * @param string|int $name
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Button.php]: Strange to allow int btw

File: src/Symfony/Component/Form/ButtonBuilder.php (hunk 1 above)
Comment (xabbuh) [reply src/Symfony/Component/Form/ButtonBuilder.php]: I think it should be safe to add the type hint here. The autoloading will only happen for listeners that are registered on forms that are actually used during the current request. And usually you shouldn't have dozens of listeners on a single form.

File: src/Symfony/Component/Form/Button.php (hunk 6 above)
Comment (xabbuh) [reply src/Symfony/Component/Form/Button.php]: This by the way looks like something that should happen anymore thanks to the changes to the `ButtonBuilder` in 4.3 IIRC.

**Comment by kaznovac** [comment]
should we typehint `\Symfony\Component\Form\FormTypeGuesserInterface` also?
(asking as i looked in typehinting doctrine bridge)

File: src/Symfony/Component/Form/FormBuilderInterface.php [hunk 9]
```diff
@@ -29,23 +29,23 @@ interface FormBuilderInterface extends \Traversable, \Countable, FormConfigBuild
...
      * @return self
      */
-    public function add($child, $type = null, array $options = []);
+    public function add($child, string $type = null, array $options = []);
 
     /**
      * Creates a form builder.
      *
-     * @param string      $name    The name of the form or the name of the property
+     * @param string|int  $name    The name of the form or the name of the property
```
Comment (kaznovac) [inline src/Symfony/Component/Form/FormBuilderInterface.php]: int feels a lot strange here:

look at this code:

1. this is FormBuilder implementation of create 
```php
    public function create($name, $type = null, array $options = [])
    {
        if ($this->locked) {
            throw new BadMethodCallException('FormBuilder methods cannot be accessed anymore once the builder is turned into a FormConfigInterface instance.');
        }

        if (null === $type && null === $this->getDataClass()) {
            $type = 'Symfony\Component\Form\Extension\Core\Type\TextType';
        }

        if (null !== $type) {
            return $this->getFormFactory()->createNamedBuilder($name, $type, null, $options);
        }

        return $this->getFormFactory()->createBuilderForProperty($this->getDataClass(), $name, null, $options);
    }
```

2. and this is `FormFactoryInterface::createBuilerForProperty`
```php
    /**
     * Returns a form builder for a property of a class.
     *
     * If any of the 'required' and type options can be guessed,
     * and are not provided in the options argument, the guessed value is used.
     *
     * @param string $class    The fully qualified class name
     * @param string $property The name of the property to guess for
     * @param mixed  $data     The initial data
     * @param array  $options  The options for the builder
     *
     * @return FormBuilderInterface The form builder named after the property
     *
     * @throws \Symfony\Component\OptionsResolver\Exception\InvalidOptionsException if any given option is not applicable to the form type
     */
    public function createBuilderForProperty($class, $property, $data = null, array $options = []);
```

3. and this is it's implementation in `FormBuilder`
```php
    public function createBuilderForProperty($class, $property, $data = null, array $options = [])
    {
        if (null === $guesser = $this->registry->getTypeGuesser()) {
            return $this->createNamedBuilder($property, 'Symfony\Component\Form\Extension\Core\Type\TextType', $data, $options);
        }

        $typeGuess = $guesser->guessType($class, $property);
        $maxLengthGuess = $guesser->guessMaxLength($class, $property);
        $requiredGuess = $guesser->guessRequired($class, $property);
        $patternGuess = $guesser->guessPattern($class, $property);

        $type = $typeGuess ? $typeGuess->getType() : 'Symfony\Component\Form\Extension\Core\Type\TextType';

        $maxLength = $maxLengthGuess ? $maxLengthGuess->getValue() : null;
        $pattern = $patternGuess ? $patternGuess->getValue() : null;

        if (null !== $pattern) {
            $options = array_replace_recursive(['attr' => ['pattern' => $pattern]], $options);
        }

        if (null !== $maxLength) {
            $options = array_replace_recursive(['attr' => ['maxlength' => $maxLength]], $options);
        }

        if ($requiredGuess) {
            $options = array_merge(['required' => $requiredGuess->getValue()], $options);
        }

        // user options may override guessed options
        if ($typeGuess) {
            $attrs = [];
            $typeGuessOptions = $typeGuess->getOptions();
            if (isset($typeGuessOptions['attr']) && isset($options['attr'])) {
                $attrs = ['attr' => array_merge($typeGuessOptions['attr'], $options['attr'])];
            }

            $options = array_merge($typeGuessOptions, $options, $attrs);
        }

        return $this->createNamedBuilder($property, $type, $data, $options);
    }
```

4. note that it uses `FormTypeGuesserInterface::guessType`
```php
    /**
     * Returns a field guess for a property name of a class.
     *
     * @param string $class    The fully qualified class name
     * @param string $property The name of the property to guess for
     *
     * @return Guess\TypeGuess|null A guess for the field's type and options
     */
    public function guessType($class, $property);
```

this leads to that int can be a property name :|

Comment (vudaltsov) [reply src/Symfony/Component/Form/FormBuilderInterface.php]: We already discussed the issue and we'll do strings only version, see https://github.com/symfony/symfony/pull/32400. 

Nevertheless, using `int` in both Forms and Builders is possible because the name later is always converted to `string` in https://github.com/symfony/symfony/blob/4.3/src/Symfony/Component/Form/FormConfigBuilder.php#L126. Btw, internally Symfony does pass ints to all these methods, see tests covering these cases. You can also try `$builder->add(10)` in your project — the resulting form will have `'10'`, so it works. That is why my idea is not nonsense.

But I agree that if we design the component from scratch, we would allow only strings. So that's what we want to achieve in 5.0.

**Comment by vudaltsov** [comment]
I will continue to work on the issue soon. `Symfony\Component\Form\FormTypeGuesserInterface` should be typehinted in this PR, I guess.

**Comment by derrabus** [comment]
> I will continue to work on the issue soon. `Symfony\Component\Form\FormTypeGuesserInterface` should be typehinted in this PR, I guess.

Yes. If you also update the corresponding implementation in the doctrine bridge, remember to update the composer.json file of the bridge accordingly.

**Comment by nicolas-grekas** [comment]
Good for rebase.

File: src/Symfony/Component/Form/FormFactoryInterface.php [hunk 10]
```diff
@@ -63,34 +61,32 @@ public function createNamed($name, $type = 'Symfony\Component\Form\Extension\Cor
...
-    public function createBuilder($type = 'Symfony\Component\Form\Extension\Core\Type\FormType', $data = null, array $options = []);
+    public function createBuilder(string $type = 'Symfony\Component\Form\Extension\Core\Type\FormType', $data = null, array $options = []);
 
     /**
      * Returns a form builder.
      *
      * @param string|int $name    The name of the form
-     * @param string     $type    The type of the form
      * @param mixed      $data    The initial data
      * @param array      $options The options
```
Comment (Tobion) [inline src/Symfony/Component/Form/FormFactoryInterface.php]: you can remove those param phpdocs

File: src/Symfony/Component/Form/FormBuilder.php [hunk 11]
```diff
@@ -66,10 +66,6 @@ public function add($child, $type = null, array $options = [])
             throw new UnexpectedTypeException($child, 'string, integer or Symfony\Component\Form\FormBuilderInterface');
         }
 
-        if (null !== $type && !\is_string($type) && !$type instanceof FormTypeInterface) {
```
Comment (Tobion) [inline src/Symfony/Component/Form/FormBuilder.php]: 👍 but we should fix this in lower branches because FormTypeInterface is indeed not a valid value here

File: src/Symfony/Component/Form/Form.php [hunk 12]
```diff
@@ -831,7 +831,7 @@ public function all()
     /**
      * {@inheritdoc}
      */
-    public function add($child, $type = null, array $options = [])
+    public function add($child, string $type = null, array $options = [])
```
Comment (Tobion) [inline src/Symfony/Component/Form/Form.php]: $type is also referencing `FormTypeInterface` in code which can be removed and should be fixed in 4.4

File: src/Symfony/Component/Form/FormBuilder.php (hunk 11 above)
Comment (Tobion) [reply src/Symfony/Component/Form/FormBuilder.php]: #32934

File: src/Symfony/Component/Form/Form.php (hunk 12 above)
Comment (Tobion) [reply src/Symfony/Component/Form/Form.php]: #32934

File: src/Symfony/Component/Form/ChoiceList/Factory/ChoiceListFactoryInterface.php (hunk 3 above)
Comment (vudaltsov) [reply src/Symfony/Component/Form/ChoiceList/Factory/ChoiceListFactoryInterface.php]: We cannot do this, because then `PropertyAccessDecorator` will not work. It accepts `string|PropertyPath` as well. I will do a PR to fix these phpdocs.

**Comment by xabbuh** [comment]
@vudaltsov Can you please rebase here? :)

File: src/Symfony/Bridge/Twig/Form/TwigRendererEngine.php [hunk 13]
```diff
@@ -40,7 +40,7 @@ public function __construct(array $defaultThemes, Environment $environment)
     /**
      * {@inheritdoc}
      */
-    public function renderBlock(FormView $view, $resource, string $blockName, array $variables = [])
```
Comment (derrabus) [inline src/Symfony/Bridge/Twig/Form/TwigRendererEngine.php]: Why do we remove the parameter type here?

Comment (vudaltsov) [reply src/Symfony/Bridge/Twig/Form/TwigRendererEngine.php]: just noticed, fixing :)
i was fixing a test and messed up here :(

**Review by derrabus — APPROVED** [review]

File: src/Symfony/Component/Form/ChoiceList/View/ChoiceView.php [hunk 14]
```diff
@@ -35,7 +35,7 @@ class ChoiceView
      * @param string $label The label displayed to humans
```
Comment (kaznovac) [inline src/Symfony/Component/Form/ChoiceList/View/ChoiceView.php]: could we state other (than string) possible types in doc?

Comment (vudaltsov) [reply src/Symfony/Component/Form/ChoiceList/View/ChoiceView.php]: See https://github.com/symfony/symfony/pull/32935

File: src/Symfony/Component/Form/Button.php [hunk 15]
```diff
@@ -148,7 +146,7 @@ public function add($child, $type = null, array $options = [])
      *
      * @throws BadMethodCallException
      */
-    public function get($name)
+    public function get(string $name)
```
Comment (Tobion) [inline src/Symfony/Component/Form/Button.php]: you can remove the param phpdoc

Comment (Tobion) [reply src/Symfony/Component/Form/Button.php]: same in the other methods like `has`

File: src/Symfony/Component/Form/Extension/Validator/ViolationMapper/MappingRule.php [hunk 16]
```diff
@@ -48,7 +48,7 @@ public function getOrigin()
      *
      * @return FormInterface|null The mapped form or null
      */
-    public function match($propertyPath)
+    public function match(string $propertyPath)
```
Comment (Tobion) [inline src/Symfony/Component/Form/Extension/Validator/ViolationMapper/MappingRule.php]: param phpdoc can be removed

File: src/Symfony/Component/Form/Extension/Validator/ViolationMapper/MappingRule.php [hunk 17]
```diff
@@ -62,7 +62,7 @@ public function match($propertyPath)
      *
      * @return bool Whether the property path is a prefix of the rule or not
      */
-    public function isPrefix($propertyPath)
+    public function isPrefix(string $propertyPath)
```
Comment (Tobion) [inline src/Symfony/Component/Form/Extension/Validator/ViolationMapper/MappingRule.php]: param phpdoc can be removed

File: src/Symfony/Component/Form/ChoiceList/Factory/ChoiceListFactoryInterface.php (hunk 3 above)
Comment (Tobion) [reply src/Symfony/Component/Form/ChoiceList/Factory/ChoiceListFactoryInterface.php]: `PropertyAccessDecorator` can remove the typehint on `$value` so it accepts more. It should be possible to add the types in the interface as implementations can widen them.

**Review by Tobion — CHANGES_REQUESTED** [review]
https://github.com/symfony/symfony/pull/32237#discussion_r298389207

File: src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php [hunk 18]
```diff
@@ -81,7 +81,7 @@ public function getDecoratedFactory()
     /**
      * {@inheritdoc}
      */
-    public function createListFromChoices($choices, $value = null)
+    public function createListFromChoices(iterable $choices, callable $value = null)
```
Comment (Tobion) [inline src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php]: you can't add the callable type to the caching decorator. otherwise we can't decorate the `PropertyAccessDecorator`

Comment (vudaltsov) [reply src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php]: Oh, yeah, missed that :(

**Comment by Tobion** [comment]
Tests are failing. The $label of createView can also be `false`, see `\Symfony\Component\Form\ChoiceList\Factory\DefaultChoiceListFactory::addChoiceView`

**Comment by vudaltsov** [comment]
@Tobion , I know that, that's exactly why I fixed it in https://github.com/symfony/symfony/pull/32935, see https://github.com/symfony/symfony/pull/32237#discussion_r310535785. I propose waiting for the merge and then I will rebase and squash everything, ok?

**Comment by Tobion** [comment]
If you know, then #32935 does not fix this completely. https://github.com/vudaltsov/symfony/blob/360711ce4e3b83251e7fe9dac958dc4f4246c0b6/src/Symfony/Component/Form/ChoiceList/Factory/ChoiceListFactoryInterface.php#L81 can also be `false`. So the phpdocs are still wrong.

**Comment by nicolas-grekas** [comment]
(merged up)

**Comment by vudaltsov** [comment]
@Tobion , got it, thank you! 

I now need some time to distract from Forms before the final fight in this PR and relevant changes in other branches :) 
I hope to finish tonight or tomorrow.

**Comment by Tobion** [comment]
#32955 should solve this problem.

**Comment by nicolas-grekas** [comment]
up for a rebase?

**Comment by vudaltsov** [comment]
@nicolas-grekas , sure, tonight I will take over https://github.com/symfony/symfony/pull/32955 and then we can safely add `?callable` to `$label`. But I will rebase this branch today as well.

**Comment by vudaltsov** [comment]
@derrabus , should I revert changes to `OptionsResolverWrapper extends OptionsResolver` and `ViolationPath implements PropertyPathInterface` to avoid bumping up the versions of the required packages?

File: src/Symfony/Component/Form/FormRendererEngineInterface.php [hunk 19]
```diff
@@ -128,7 +128,6 @@ public function getResourceHierarchyLevel(FormView $view, array $blockNameHierar
      *
      * @param FormView $view      The view to render
      * @param mixed    $resource  The renderer resource
-     * @param string   $blockName The name of the block to render
```
Comment (xabbuh) [inline src/Symfony/Component/Form/FormRendererEngineInterface.php]: Why did you remove this one but not the others?

File: src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php [hunk 20]
```diff
@@ -81,7 +81,7 @@ public function getDecoratedFactory()
     /**
      * {@inheritdoc}
      */
-    public function createListFromChoices($choices, $value = null)
+    public function createListFromChoices(iterable $choices, $value = null)
```
Comment (xabbuh) [inline src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php]: `callable $value = null`?

File: src/Symfony/Component/Form/DataMapperInterface.php [hunk 21]
```diff
@@ -22,12 +22,12 @@ interface DataMapperInterface
      * The method is responsible for calling {@link FormInterface::setData()}
      * on the children of compound forms, defining their underlying model data.
      *
-     * @param mixed                        $viewData View data of the compound form being initialized
-     * @param FormInterface[]|\Traversable $forms    A list of {@link FormInterface} instances
+     * @param mixed                    $viewData View data of the compound form being initialized
+     * @param FormInterface[]|iterable $forms    A list of {@link FormInterface} instances
```
Comment (xabbuh) [inline src/Symfony/Component/Form/DataMapperInterface.php]: Would `iterable<FormInterface>` be a correct notation here?

File: src/Symfony/Component/Form/DataMapperInterface.php [hunk 22]
```diff
@@ -52,11 +52,11 @@ public function mapDataToForms($viewData, $forms);
      * The model data can be an array or an object, so this second argument is always passed
      * by reference.
      *
-     * @param FormInterface[]|\Traversable $forms    A list of {@link FormInterface} instances
-     * @param mixed                        $viewData The compound form's view data that get mapped
-     *                                               its children model data
+     * @param FormInterface[]|iterable $forms    A list of {@link FormInterface} instances
```
Comment (xabbuh) [inline src/Symfony/Component/Form/DataMapperInterface.php]: Would `iterable<FormInterface>` be a correct notation here?

File: src/Symfony/Component/Form/DataMapperInterface.php (hunk 21 above)
Comment (vudaltsov) [reply src/Symfony/Component/Form/DataMapperInterface.php]: it's a correct notation, but do we use it in Symfony?

File: src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php (hunk 20 above)
Comment (vudaltsov) [reply src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php]: we cannot do this, because then `string|PropertyPathInterface $value` will not be acceptable in this caching layer

Comment (vudaltsov) [reply src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php]: see https://github.com/symfony/symfony/pull/32237#discussion_r310611265

File: src/Symfony/Component/Form/ChoiceList/View/ChoiceGroupView.php [hunk 23]
```diff
@@ -27,7 +27,7 @@ class ChoiceGroupView implements \IteratorAggregate
      * @param string                         $label   The label of the group
```
[the remark this answers is not in the document]
Comment (vudaltsov) [reply src/Symfony/Component/Form/ChoiceList/View/ChoiceGroupView.php]: done

File: src/Symfony/Component/Form/FormRendererEngineInterface.php (hunk 19 above)
Comment (vudaltsov) [reply src/Symfony/Component/Form/FormRendererEngineInterface.php]: Think it happened when rebasing. I added `string` declaration myself and then someone added it too without removing the phpdoc, so only phpdoc removal left here.
Should I remove `$view` in another PR targeting `3.4`? Or just revert my change in this class?

**Comment by xabbuh** [comment]
We need to update the minimum required version for the `symfony/property-access` package to 5.0.

**Comment by vudaltsov** [comment]
@xabbuh , for `options-resolver` as well, right? Since we have `OptionsResolverWrapper`. Done

File: src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php [hunk 24]
```diff
@@ -29,23 +29,23 @@ class DefaultChoiceListFactory implements ChoiceListFactoryInterface
...
+    public function createListFromLoader(ChoiceLoaderInterface $loader, callable $value = null)
     {
         return new LazyChoiceList($loader, $value);
     }
 
     /**
      * {@inheritdoc}
      */
-    public function createView(ChoiceListInterface $list, $preferredChoices = null, $label = null, $index = null, $groupBy = null, $attr = null)
+    public function createView(ChoiceListInterface $list, $preferredChoices = null, callable $label = null, callable $index = null, callable $groupBy = null, $attr = null)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php]: this change breaks TwigBundle, which passes a `false` label:
> TypeError: Argument 3 passed to Symfony\Component\Form\ChoiceList\Factory\DefaultChoiceListFactory::createView() must be callable or null, boolean given, called in /home/travis/build/symfony/symfony/src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php on line 193

Comment (Tobion) [reply src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php]: See #33074

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @vudaltsov.
